### ISSUE TITLE
data is not parsed correctly

### ISSUE THREAD

**Issue by bensLine** [emqtt/qmqtt#72]
Hey,

qmqtt crashes for me regularly in the following method
```
quint16 Frame::readInt()
{
    char msb = _data.at(0);
    char lsb = _data.at(1);
    _data.remove(0, 2);
    return (msb << 8) + lsb;
}
```
the problem is that the `_data` is empty but there is no check for it so the assertion in .at() is triggered.
However, after a bit of debugging I think the problem happens before.

in `void QMQTT::Network::onSocketReadReady()` a Frame is created with the data
`"8980,"x":-0.236530,"y":0.748594,"z":-0.022885,"ts"`
Then this data is parsed in `void QMQTT::ClientPrivate::onNetworkReceived(const QMQTT::Frame& frm)` as a message (PUBLISH case) with the topic `"80,"x":-0.236530,"y":0.748594,"z":-0.022885,"ts"` as the `89` was interpreted as topic length and hence the frame data is empty. In the next step the crash occurs as the `mid` cannot be read from the empty data.

Obviously the data for the Frame in in `void QMQTT::Network::onSocketReadReady()` is incomplete. I actually send a JSON string like: `"{ "acc": { "x":1, "y":1, "z":1,"ts":1} }"` to a topic. Anyway, I did not investigate any further and I'm not aware how the mqtt data is send and received.

Any ideas how it comes to this behavior?

**Comment by mwallnoefer** [comment]
Does your broker generate standard-compliant Mqtt messages? (spec: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html)

Otherwise could you please provide a sample raw Mqtt message (socket) so that we can find out where our parser fails?

**Comment by bensLine** [comment]
Thanks for the quick reply!
Yes that should be fine, I'm using mosquitto. This bug is not always happening though. I can receive messages with qmqtt and they get parsed correctly but usually after 30s qmqtt crashes due to the mentioned error. I'm sending data at 10Hz btw.

With MQTT.fx for example the message is parsed correctly.
I recorded the failing message with Wireshark, you can see it bellow but you can download the logs [here](https://drive.google.com/folderview?id=0B6gg8-WZU6nKZE9vNHZiWDlWak0&usp=sharing) too. The first MQTT message in the logs was parsed correctly, the second MQTT message in the log failed and lead to the following data data in `void QMQTT::Network::onSocketReadReady()`:
`"x":0.000000,"y":0.000000,"z":0.000000,"ts":0},"gyro":{"x":0.019196,"y":-0.007464,"z":-0.017906,"ts":"1467454928364406738"},"` with header `58`

failing message:
```
0000  0c 8b fd cc b3 bb 44 74 6c 83 b7 67 08 00 45 00   ......Dtl..g..E.
0010  04 66 b3 ba 40 00 40 06 00 4e c0 a8 00 8c c0 a8   .f..@.@..N......
0020  00 ad 8b 29 07 5b b9 02 87 9a 3e f8 00 a8 80 18   ...).[....>.....
0030  05 59 21 66 00 00 01 01 08 0a 00 09 83 01 18 f7   .Y!f............
0040  23 75 30 af 08 00 15 62 6c 61 63 6b 62 6f 61 72   #u0....blackboar
0050  64 2f 73 65 6e 73 6f 72 44 61 74 61 7b 22 74 79   d/sensorData{"ty
0060  70 65 22 3a 22 53 65 6e 73 6f 72 44 61 74 61 4c   pe":"SensorDataL
0070  69 73 74 22 2c 22 73 65 6e 73 6f 72 5f 64 61 74   ist","sensor_dat
0080  61 22 3a 5b 0a 7b 22 74 79 70 65 22 3a 22 53 65   a":[.{"type":"Se
0090  6e 73 6f 72 44 61 74 61 22 2c 22 61 63 63 22 3a   nsorData","acc":
00a0  7b 22 78 22 3a 30 2e 30 30 30 30 30 30 2c 22 79   {"x":0.000000,"y
00b0  22 3a 30 2e 30 30 30 30 30 30 2c 22 7a 22 3a 30   ":0.000000,"z":0
00c0  2e 30 30 30 30 30 30 2c 22 74 73 22 3a 30 7d 2c   .000000,"ts":0},
00d0  22 67 79 72 6f 22 3a 7b 22 78 22 3a 30 2e 30 31   "gyro":{"x":0.01
00e0  32 36 35 39 2c 22 79 22 3a 2d 30 2e 30 30 36 32   2659,"y":-0.0062
00f0  33 37 2c 22 7a 22 3a 2d 30 2e 30 30 38 33 39 31   37,"z":-0.008391
0100  2c 22 74 73 22 3a 31 34 36 37 34 35 34 39 32 38   ,"ts":1467454928
0110  33 33 34 34 36 38 39 39 34 7d 2c 22 72 6f 74 22   334468994},"rot"
0120  3a 7b 22 77 22 3a 2d 30 2e 36 31 32 31 36 34 2c   :{"w":-0.612164,
0130  22 78 22 3a 30 2e 34 30 35 33 33 35 2c 22 79 22   "x":0.405335,"y"
0140  3a 2d 30 2e 36 33 38 34 38 30 2c 22 7a 22 3a 30   :-0.638480,"z":0
0150  2e 32 33 30 38 37 33 2c 22 74 73 22 3a 31 34 36   .230873,"ts":146
0160  37 34 35 34 39 32 38 33 34 34 34 37 38 37 36 30   7454928344478760
0170  7d 7d 2c 0a 7b 22 74 79 70 65 22 3a 22 53 65 6e   }},.{"type":"Sen
0180  73 6f 72 44 61 74 61 22 2c 22 61 63 63 22 3a 7b   sorData","acc":{
0190  22 78 22 3a 30 2e 30 30 30 30 30 30 2c 22 79 22   "x":0.000000,"y"
01a0  3a 30 2e 30 30 30 30 30 30 2c 22 7a 22 3a 30 2e   :0.000000,"z":0.
01b0  30 30 30 30 30 30 2c 22 74 73 22 3a 30 7d 2c 22   000000,"ts":0},"
01c0  67 79 72 6f 22 3a 7b 22 78 22 3a 30 2e 30 31 39   gyro":{"x":0.019
01d0  31 39 36 2c 22 79 22 3a 2d 30 2e 30 30 37 34 36   196,"y":-0.00746
01e0  34 2c 22 7a 22 3a 2d 30 2e 30 31 37 39 30 36 2c   4,"z":-0.017906,
01f0  22 74 73 22 3a 31 34 36 37 34 35 34 39 32 38 33   "ts":14674549283
0200  36 34 34 30 36 37 33 38 7d 2c 22 72 6f 74 22 3a   64406738},"rot":
0210  7b 22 77 22 3a 2d 30 2e 36 31 32 31 33 32 2c 22   {"w":-0.612132,"
0220  78 22 3a 30 2e 34 30 35 34 32 33 2c 22 79 22 3a   x":0.405423,"y":
0230  2d 30 2e 36 33 38 31 39 30 2c 22 7a 22 3a 30 2e   -0.638190,"z":0.
0240  32 33 31 36 30 33 2c 22 74 73 22 3a 31 34 36 37   231603,"ts":1467
0250  34 35 34 39 32 38 33 38 34 32 31 32 36 34 36 7d   454928384212646}
0260  7d 2c 0a 7b 22 74 79 70 65 22 3a 22 53 65 6e 73   },.{"type":"Sens
0270  6f 72 44 61 74 61 22 2c 22 61 63 63 22 3a 7b 22   orData","acc":{"
0280  78 22 3a 30 2e 30 30 30 30 30 30 2c 22 79 22 3a   x":0.000000,"y":
0290  30 2e 30 30 30 30 30 30 2c 22 7a 22 3a 30 2e 30   0.000000,"z":0.0
02a0  30 30 30 30 30 2c 22 74 73 22 3a 30 7d 2c 22 67   00000,"ts":0},"g
02b0  79 72 6f 22 3a 7b 22 78 22 3a 2d 30 2e 30 30 34   yro":{"x":-0.004
02c0  32 32 32 2c 22 79 22 3a 2d 30 2e 30 30 30 39 34   222,"y":-0.00094
02d0  33 2c 22 7a 22 3a 30 2e 30 30 36 31 36 37 2c 22   3,"z":0.006167,"
02e0  74 73 22 3a 31 34 36 37 34 35 34 39 32 38 33 39   ts":146745492839
02f0  34 35 32 37 35 38 38 7d 2c 22 72 6f 74 22 3a 7b   4527588},"rot":{
0300  22 77 22 3a 2d 30 2e 36 31 32 31 33 39 2c 22 78   "w":-0.612139,"x
0310  22 3a 30 2e 34 30 35 34 30 35 2c 22 79 22 3a 2d   ":0.405405,"y":-
0320  30 2e 36 33 38 32 34 35 2c 22 7a 22 3a 30 2e 32   0.638245,"z":0.2
0330  33 31 34 36 33 2c 22 74 73 22 3a 31 34 36 37 34   31463,"ts":14674
0340  35 34 39 32 38 34 30 34 34 37 36 33 31 38 7d 7d   54928404476318}}
0350  2c 0a 7b 22 74 79 70 65 22 3a 22 53 65 6e 73 6f   ,.{"type":"Senso
0360  72 44 61 74 61 22 2c 22 61 63 63 22 3a 7b 22 78   rData","acc":{"x
0370  22 3a 30 2e 30 30 30 30 30 30 2c 22 79 22 3a 30   ":0.000000,"y":0
0380  2e 30 30 30 30 30 30 2c 22 7a 22 3a 30 2e 30 30   .000000,"z":0.00
0390  30 30 30 30 2c 22 74 73 22 3a 30 7d 2c 22 67 79   0000,"ts":0},"gy
03a0  72 6f 22 3a 7b 22 78 22 3a 2d 30 2e 30 32 30 31   ro":{"x":-0.0201
03b0  30 31 2c 22 79 22 3a 30 2e 30 31 32 39 37 32 2c   01,"y":0.012972,
03c0  22 7a 22 3a 30 2e 30 31 35 38 32 38 2c 22 74 73   "z":0.015828,"ts
03d0  22 3a 31 34 36 37 34 35 34 39 32 38 34 32 34 34   ":14674549284244
03e0  36 35 33 33 32 7d 2c 22 72 6f 74 22 3a 7b 22 77   65332},"rot":{"w
03f0  22 3a 2d 30 2e 36 31 31 37 36 36 2c 22 78 22 3a   ":-0.611766,"x":
0400  30 2e 34 30 35 34 33 35 2c 22 79 22 3a 2d 30 2e   0.405435,"y":-0.
0410  36 33 38 35 32 35 2c 22 7a 22 3a 30 2e 32 33 31   638525,"z":0.231
0420  36 32 37 2c 22 74 73 22 3a 31 34 36 37 34 35 34   627,"ts":1467454
0430  39 32 38 34 33 34 35 30 35 36 31 35 7d 7d 5d 2c   928434505615}}],
0440  0a 22 63 6f 75 6e 74 22 3a 34 2c 0a 22 6c 61 73   ."count":4,."las
0450  74 5f 74 69 6d 65 73 74 61 6d 70 22 3a 22 31 34   t_timestamp":"14
0460  36 37 34 35 34 39 32 38 34 33 34 35 30 35 36 31   6745492843450561
0470  35 22 0a 7d                                       5".}
```

**Comment by ejvr** [comment]
This issue looks a lot like the problems I had a few months ago. My issue was fixed by commit 7b805464cadc2116ec6ef03dcc02e631e61a1e7e which is part of the master branch. Are you using the latest version of the master branch?

By the way: there's also a problem in the Frame::readInt() function: on most systems char's are signed by default, so the lsb variable may become negative, which will screw up the return value of the function whenever the most significant bit of the LSB is 1. I'll try to create a pull request for this repository today.

**Comment by bensLine** [comment]
oh, my bad! Indeed, it seems like this was the problem :) Thanks!
